desc/coils.py

```diff
@@ -6,16 +6,107 @@
+    """Biot-Savart law for filamentary coils following [1].
+
+    The coil is approximated by a series of straight line segments
+    and an analytic expression is used to evaluate the field from each
+    segment.
+
+    Parameters
+    ----------
+    eval_pts : array-like shape(n,3)
+        Evaluation points in cartesian coordinates
+    coil_pts_start, coil_pts_end : array-like shape(m,3)
+        Points in cartesian space defining the start and end of each segment.
+        Should be a closed curve, such that coil_pts_start[0] == coil_pts_end[-1]
+        though this is not checked.
+    current : float
+        Current through the coil (in Amps).
+
+    Returns
+    -------
+    B : ndarray, shape(n,3)
+        magnetic field in cartesian components at specified points
+
+    [1] Hanson & Hirshman, "Compact expressions for the Biot-Savart
+    fields of a filamentary segment" (2002)
+    """
+    d_vec = coil_pts_end - coil_pts_start
+    L = jnp.linalg.norm(d_vec, axis=-1)
+
+    Ri_vec = eval_pts[jnp.newaxis, :] - coil_pts_start[:, jnp.newaxis, :]
+    Ri = jnp.linalg.norm(Ri_vec, axis=-1)
+    Rf = jnp.linalg.norm(
+        eval_pts[jnp.newaxis, :] - coil_pts_end[:, jnp.newaxis, :], axis=-1
+    )
+    Ri_p_Rf = Ri + Rf
+
+    # 1.0e-7 == mu_0/(4 pi)
+    B_mag = (
+        1.0e-7
+        * current
+        * 2.0
+        * Ri_p_Rf
+        / (Ri * Rf * (Ri_p_Rf * Ri_p_Rf - (L * L)[:, jnp.newaxis]))
+    )
+
+    # cross product of L*hat(eps)==d_vec with Ri_vec, scaled by B_mag
+    vec = jnp.cross(d_vec[:, jnp.newaxis, :], Ri_vec, axis=-1)
+    B = jnp.sum(B_mag[:, :, jnp.newaxis] * vec, axis=0)
+    return B
+
+
+def biot_savart_quad(eval_pts, coil_pts, tangents, current):
+    """Biot-Savart law for filamentary coil using numerical quadrature.
+
+    Parameters
+    ----------
+    eval_pts : array-like shape(n,3)
+        Evaluation points in cartesian coordinates
+    coil_pts : array-like shape(m,3)
+        Points in cartesian space defining coil
+    tangents : array-like, shape(m,3)
+        Tangent vectors to the coil at coil_pts. If the curve is given
+        by x(s) with curve parameter s, coil_pts = x, tangents = dx/ds*ds where
+        ds is the spacing between points.
+    current : float
+        Current through the coil (in Amps).
+
+    Returns
+    -------
+    B : ndarray, shape(n,3)
+        magnetic field in cartesian components at specified points
+
+    Notes
+    -----
+    This method does not give curl(B) == 0 exactly. The error in the curl
+    scales the same as the error in B itself, so will only be zero when fully
+    converged. However in practice, for smooth curves described by Fourier series,
+    this method converges exponentially in the number of coil points.
+    """
+    dl = tangents
+    R_vec = eval_pts[jnp.newaxis, :] - coil_pts[:, jnp.newaxis, :]
+    R_mag = jnp.linalg.norm(R_vec, axis=-1)
+
+    vec = jnp.cross(dl[:, jnp.newaxis, :], R_vec, axis=-1)
+    denom = R_mag * R_mag * R_mag
```

Comment:
Is this any better than `R_mag**3`? The latter is more readable in my opinion.

Reply:
in regular numpy powers that aren't `2` get passed to a floating point `pow` routine which is relatively slow (because its designed for stuff like `x**2.12039850984509324`) while stuff like `x**2` gets evaluated as `x*x` which is super fast.

It looks like jax is smart enough to evaluate `x**3` as `x*x*x` if its jitted so I'll replace with that.